Sources/SwiftDocC/Infrastructure/DocumentationContext.swift

```diff
@@ -477,57 +436,54 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
+
+                // Avoid any synchronization overhead if there are no references to add.
+                guard !externalLinksCollector.collectedExternalReferences.isEmpty else { return }
                 
                 // Add the link pairs to `collectedExternalLinks`.
-                results.append(contentsOf:
-                    externalLinksCollector.collectedExternalReferences.map { unresolved -> ExternalLinkResult in
-                        return ExternalLinkResult(unresolved: unresolved, targetLanguage: object.reference.sourceLanguage)
+                collectedExternalLinks.sync {
+                    for (bundleID, collectedLinks) in externalLinksCollector.collectedExternalReferences {
+                        $0[bundleID, default: []].formUnion(collectedLinks)
                     }
-                )
+                }
             }
         }
         
-        try Set(collectedExternalLinks).compactMap { externalLink -> (url: ValidatedURL, resolved: TopicReferenceResolutionResult)? in
-            guard let referenceBundleIdentifier = externalLink.unresolved.topicURL.components.host else {
-                assertionFailure("Should not hit this code path, url is verified to have an external bundle id.")
-                return nil
+        for (bundleID, collectedLinks) in collectedExternalLinks.sync({ $0 }) {
+            guard let externalResolver = externalDocumentationSources[bundleID] else {
+                continue
             }
-            
-            if let externalResolver = externalReferenceResolvers[referenceBundleIdentifier] {
-                let reference = externalResolver.resolve(.unresolved(externalLink.unresolved), sourceLanguage: externalLink.targetLanguage)
-                if case .success(let resolvedReference) = reference {
+            for externalLink in collectedLinks {
+                let unresolvedURL = externalLink.topicURL
+                let result = externalResolver.resolve(.unresolved(externalLink))
+                externallyResolvedLinks[unresolvedURL] = result
+                
+                if case .success(let resolvedReference) = result {
                     // Add the resolved entity to the documentation cache.
-                    if let externallyResolvedNode = try externalEntity(with: resolvedReference) {
-                        documentationCache[resolvedReference] = externallyResolvedNode
+                    if let externallyResolvedNode = externalEntity(with: resolvedReference) {
+                        externalCache[resolvedReference] = externallyResolvedNode
+                    }
+                    if unresolvedURL.absoluteString != resolvedReference.absoluteString,
+                       let resolvedURL = ValidatedURL(resolvedReference.url) {
+                        // If the resolved reference has a different URL than the authored link, cache both URLs so we can resolve both unresolved and resolved references.
```

Comment:
When would the resolved reference have a different URL than the authored link?

Reply:
The two main examples would be if the external reference was:
- a redirected article and the authored link was the old URL
- a symbol with multiple language representations and the authored link wasn't the canonical URL

Reply:
I added this information in https://github.com/apple/swift-docc/pull/806/commits/91254ddbe6350e8105e28a59276df74973b72c4c